app/src/main/java/com/b_lam/resplash/ui/photo/detail/PhotoDetailActivity.kt

```diff
@@ -310,22 +342,30 @@ class PhotoDetailActivity :
 
     private fun applyWallpaper(uri: Uri) {
         try {
-            startActivity(WallpaperManager.getInstance(this).getCropAndSetWallpaperIntent(uri))
+            startActivity(wallpaperManager.getCropAndSetWallpaperIntent(uri))
         } catch (e: IllegalArgumentException) {
-            var bitmap: Bitmap? = null
-            try {
-                bitmap = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                    ImageDecoder.decodeBitmap(ImageDecoder.createSource(contentResolver, uri))
-                } else {
-                    MediaStore.Images.Media.getBitmap(contentResolver, uri)
-                }
-                WallpaperManager.getInstance(applicationContext).setBitmap(bitmap)
-                toast("Wallpaper set successfully")
-            } catch (e: Exception) {
-                error("Error setting wallpaper", e)
+            viewModel.prepareBitmapFromUri(contentResolver, uri)
+            observerWallpaperBitmapResult()
+        }
+    }
+
+    private fun observerWallpaperBitmapResult() {
+        viewModel.wallpaperBitmap.observe(this) { result ->
+            if (result is Result.Success) {
+                setWallpaperWithBitmap(result.value)
+            } else {
                 toast("Failed to set wallpaper")
-            } finally {
-                bitmap?.recycle()
+            }
+        }
+    }
+
+    private fun setWallpaperWithBitmap(bitmap: Bitmap) {
+        lifecycleScope.launch {
+            withContext((Executors.newSingleThreadExecutor().asCoroutineDispatcher())) {
+                wallpaperManager.setBitmap(bitmap)
```

Comment:
`setBitmap` throws IOException in case if something goes wrong while setting wallpaper. Please handle the exception.
[Reference](https://developer.android.com/reference/android/app/WallpaperManager#setBitmap(android.graphics.Bitmap))

Reply:
Made the necessary changes.